### ISSUE TITLE
cannot import name 'is_torch_npu_available' from 'transformers'

### ISSUE THREAD

**Issue by astrowonk** [UKPLab/sentence-transformers#2895]
fresh conda environment, I conda install sentence-transformers and:

```
File /usr/local/opt/miniforge3/envs/sentencet/lib/python3.12/site-packages/sentence_transformers/cross_encoder/__init__.py:1
----> 1 from .CrossEncoder import CrossEncoder
      3 __all__ = ["CrossEncoder"]

File /usr/local/opt/miniforge3/envs/sentencet/lib/python3.12/site-packages/sentence_transformers/cross_encoder/CrossEncoder.py:12
     10 from torch.utils.data import DataLoader
     11 from tqdm.autonotebook import tqdm, trange
---> 12 from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer, is_torch_npu_available
     13 from transformers.tokenization_utils_base import BatchEncoding
     14 from transformers.utils import PushToHubMixin
```

Presumably some version compatibility issue as the import is  but … what do I need to change? older torch?  And shouldn't dependency versions be pinned so that the right things get installed on a fresh environmen?

**Comment by astrowonk** [comment]
Actually, I believe that `from transformers import is_torch_tpu_available`

is deprecated and and was taken out in 4.41.0. I think the replacement is `from transformers.utils.import_utils import is_torch_tpu_available`

**Comment by tomaarsen** [comment]
Hello!

> And shouldn't dependency versions be pinned so that the right things get installed on a fresh environmen?

Yes. However, we don't set upper bounds on dependencies under the expectation that dependencies don't make breaking changes without a major version increase. E.g. removing a function like `is_torch_npu_available` would be such an unexpected breaking change.

However, I'm struggling a bit to reproduce this. The `is_torch_npu_available` function still exists in `transformers`, even on `main`: https://github.com/huggingface/transformers/blob/85345bb439652d3f03bb4e123cef7a440f2ba95b/src/transformers/utils/import_utils.py#L640

I think the issue that you're experiencing is that your `transformers` dependency was not installed correctly, resulting in the `cannot import name 'is_torch_npu_available' from 'transformers'`. You can try to uninstall it and reinstall it to see if that helps.

- Tom Aarsen

**Comment by astrowonk** [comment]
Maybe it's a `miniforge` / conda-forge problem more so than conda. I tried again. TLDR; the import is in tranformers 4.18 but in new environment, somehow 4.17 is getting installed:


```
 conda create -n "tryagain" python=3.10
 conda install sentence-transformers 
 
```
And here is what it installed after `conda list`:

<details>
```
# packages in environment at /usr/local/opt/miniforge3/envs/tryagain:
#
# Name                    Version                   Build  Channel
brotli-python             1.1.0           py310h1253130_1    conda-forge
bzip2                     1.0.8                h99b78c6_7    conda-forge
ca-certificates           2024.7.4             hf0a4a13_0    conda-forge
certifi                   2024.7.4           pyhd8ed1ab_0    conda-forge
cffi                      1.17.0          py310hf7f7c9b_0    conda-forge
charset-normalizer        3.3.2              pyhd8ed1ab_0    conda-forge
click                     8.1.7           unix_pyh707e725_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
dataclasses               0.8                pyhc8e2a94_3    conda-forge
filelock                  3.15.4             pyhd8ed1ab_0    conda-forge
freetype                  2.12.1               hadb7bae_2    conda-forge
fsspec                    2024.6.1           pyhff2d567_0    conda-forge
gmp                       6.3.0                h7bae524_2    conda-forge
gmpy2                     2.1.5           py310h3bc658a_1    conda-forge
h2                        4.1.0              pyhd8ed1ab_0    conda-forge
hpack                     4.0.0              pyh9f0ad1d_0    conda-forge
huggingface_hub           0.24.6             pyhd8ed1ab_0    conda-forge
hyperframe                6.0.1              pyhd8ed1ab_0    conda-forge
idna                      3.7                pyhd8ed1ab_0    conda-forge
importlib-metadata        8.3.0              pyha770c72_0    conda-forge
importlib_metadata        8.3.0                hd8ed1ab_0    conda-forge
jinja2                    3.1.4              pyhd8ed1ab_0    conda-forge
joblib                    1.4.2              pyhd8ed1ab_0    conda-forge
lcms2                     2.16                 ha0e7c42_0    conda-forge
lerc                      4.0.0                h9a09cb3_0    conda-forge
libabseil                 20240116.2      cxx17_h00cdb27_1    conda-forge
libblas                   3.9.0           23_osxarm64_openblas    conda-forge
libcblas                  3.9.0           23_osxarm64_openblas    conda-forge
libcxx                    18.1.8               h5a72898_4    conda-forge
libdeflate                1.21                 h99b78c6_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libgfortran               5.0.0           13_2_0_hd922786_3    conda-forge
libgfortran5              13.2.0               hf226fd6_3    conda-forge
libjpeg-turbo             3.0.0                hb547adb_1    conda-forge
liblapack                 3.9.0           23_osxarm64_openblas    conda-forge
libopenblas               0.3.27          openmp_h517c56d_1    conda-forge
libpng                    1.6.43               h091b4b1_0    conda-forge
libprotobuf               4.25.3               hbfab5d5_0    conda-forge
libsentencepiece          0.2.0                hdceaaab_2    conda-forge
libsqlite                 3.46.0               hfb93653_0    conda-forge
libtiff                   4.6.0                hf8409c0_4    conda-forge
libtorch                  2.3.1           cpu_generic_hf1facdc_1    conda-forge
libuv                     1.48.0               h93a5062_0    conda-forge
libwebp-base              1.4.0                h93a5062_0    conda-forge
libxcb                    1.16                 hf2054a2_0    conda-forge
libzlib                   1.3.1                hfb2fe0b_1    conda-forge
llvm-openmp               18.1.8               hde57baf_1    conda-forge
markupsafe                2.1.5           py310hd125d64_0    conda-forge
mpc                       1.3.1                h91ba8db_0    conda-forge
mpfr                      4.2.1                h1cfca0a_2    conda-forge
mpmath                    1.3.0              pyhd8ed1ab_0    conda-forge
ncurses                   6.5                  hb89a1cb_0    conda-forge
networkx                  3.3                pyhd8ed1ab_1    conda-forge
nltk                      3.9.1              pyhd8ed1ab_0    conda-forge
nomkl                     1.0                  h5ca1d4c_0    conda-forge
numpy                     2.1.0           py310h52bbd9b_0    conda-forge
openjpeg                  2.5.2                h9f1df11_0    conda-forge
openssl                   3.3.1                hfb2fe0b_2    conda-forge
packaging                 24.1               pyhd8ed1ab_0    conda-forge
pillow                    10.4.0          py310h01af8b1_0    conda-forge
pip                       24.2               pyhd8ed1ab_0    conda-forge
pthread-stubs             0.4               h27ca646_1001    conda-forge
pycparser                 2.22               pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
python                    3.10.14         h2469fbe_0_cpython    conda-forge
python_abi                3.10                    5_cp310    conda-forge
pytorch                   2.3.1           cpu_generic_py310hb190f2a_1    conda-forge
pyyaml                    6.0.2           py310ha6dd24b_0    conda-forge
readline                  8.2                  h92ec313_1    conda-forge
regex                     2024.7.24       py310ha6dd24b_0    conda-forge
requests                  2.32.3             pyhd8ed1ab_0    conda-forge
sacremoses                0.0.53             pyhd8ed1ab_0    conda-forge
scikit-learn              1.5.1           py310h64e73be_0    conda-forge
scipy                     1.14.0          py310h7057308_2    conda-forge
sentence-transformers     3.0.1              pyhd8ed1ab_0    conda-forge
sentencepiece             0.2.0                hb6292c7_2    conda-forge
sentencepiece-python      0.2.0           py310h16e3829_2    conda-forge
sentencepiece-spm         0.2.0                hdceaaab_2    conda-forge
setuptools                72.1.0             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sleef                     3.6.1                hf6b88df_1    conda-forge
sympy                     1.13.2          pypyh2585a3b_103    conda-forge
threadpoolctl             3.5.0              pyhc1e730c_0    conda-forge
tk                        8.6.13               h5083fa2_1    conda-forge
tokenizers                0.20.0          py310hdeb0604_0    conda-forge
torchvision               0.18.1          cpu_py310h3a75613_1    conda-forge
tqdm                      4.66.5             pyhd8ed1ab_0    conda-forge
transformers              4.17.0             pyhd8ed1ab_0    conda-forge
typing-extensions         4.12.2               hd8ed1ab_0    conda-forge
typing_extensions         4.12.2             pyha770c72_0    conda-forge
tzdata                    2024a                h0c530f3_0    conda-forge
urllib3                   2.2.2              pyhd8ed1ab_1    conda-forge
wheel                     0.44.0             pyhd8ed1ab_0    conda-forge
xorg-libxau               1.0.11               hb547adb_0    conda-forge
xorg-libxdmcp             1.1.3                h27ca646_0    conda-forge
xz                        5.2.6                h57fd34a_0    conda-forge
yaml                      0.2.5                h3422bc3_2    conda-forge
zipp                      3.20.0             pyhd8ed1ab_0    conda-forge
zstandard                 0.23.0          py310hd9c37c8_0    conda-forge
zstd                      1.5.6                hb46c0d2_0    conda-forge
```
</details 
 
 
 and then in the REPL we're right back to:
 
 ```
 >>> from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer, is_torch_npu_available
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'is_torch_npu_available' from 'transformers' (/usr/local/opt/miniforge3/envs/tryagain/lib/python3.10/site-packages/transformers/__init__.py)
```

For whatever reason, the install of sentence-transformers with conda and miniforge is installing transformers 4.17 which is [missing the import](https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/__init__.py). It's in 4.18 though. ¯\_(ツ)_/¯ 

`conda update transformers` then solves the problem, but numpy gets downgraded to 1.26. ¯\_(ツ)_/¯  Somehow the dependency resolver is getting tripped up - I wonder if making the minimum transformer version 4.18 would help?

**Comment by cohml** [comment]
I just encountered the same error.

```shell
❯ conda create --yes --prefix ./env sentence-transformers
...
❯ conda list --prefix ./env | grep transformers |& sed "s|$HOME|<redacted>|g"
# packages in environment at <redacted>/env:
sentence-transformers     3.0.1              pyhd8ed1ab_0    conda-forge
transformers              4.17.0             pyhd8ed1ab_0    conda-forge
❯ python -q
❯ ./env/bin/python -c 'import sentence_transformers' |& sed "s|$HOME|<redacted>|g"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<redacted>/del/env/lib/python3.12/site-packages/sentence_transformers/__init__.py", line 7, in <module>
    from sentence_transformers.cross_encoder.CrossEncoder import CrossEncoder
  File "<redacted>/del/env/lib/python3.12/site-packages/sentence_transformers/cross_encoder/__init__.py", line 1, in <module>
    from .CrossEncoder import CrossEncoder
  File "<redacted>/del/env/lib/python3.12/site-packages/sentence_transformers/cross_encoder/CrossEncoder.py", line 12, in <module>
    from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer, is_torch_npu_available
ImportError: cannot import name 'is_torch_npu_available' from 'transformers' (<redacted>/del/env/lib/python3.12/site-packages/transformers/__init__.py)
❯ find ./env -type f -name '*.py' -exec grep -l is_torch_npu_available {} +
./env/lib/python3.12/site-packages/sentence_transformers/util.py
./env/lib/python3.12/site-packages/sentence_transformers/SentenceTransformer.py
./env/lib/python3.12/site-packages/sentence_transformers/cross_encoder/CrossEncoder.py
```

So in my environment, which you should be able to reproduce using the commands above, only `sentence_transformers` seems to think `is_torch_npu_available` exists. It definitely is not a part of my local `torch`:

```shell
❯ file ./env/lib/python3.12/site-packages/transformers
./env/lib/python3.12/site-packages/transformers: directory
❯ [[ -z $(find ./env/lib/python3.12/site-packages/transformers -type f -name '*.py' -exec grep is_torch_npu_available {} +) ]] && echo not found in transformers
not found in transformers
```

**Comment by tomaarsen** [comment]
Hello!

I suspect that your `transformers` version is too old for `sentence-transformers`. For context, we now require `transformers` 4.38.0 or higher: https://github.com/UKPLab/sentence-transformers/blob/dafe2b6c88c6940dffffb0c5faf2932f1461e94c/pyproject.toml#L33

Whereas `transformers` 4.17.0 is from 2.5 years ago.

---

I've just double-checked, and Sentence Transformers v3 released with a requirement of at least `transformers` 4.34.0: https://github.com/UKPLab/sentence-transformers/blob/v3.0-release/requirements.txt#L1

In short, you should be able to resolve this with:

```
pip install -U transformers
```

- Tom Aarsen

**Comment by cohml** [comment]
Thanks for the prompt reply. I did tinker a bit just now and found that pinning a newer version of `transformers` in the `conda install` did resolve the issue.

But doesn't that suggest the issue isn't with me (or you), but rather with `conda`? Is `pip` the recommended way to install `sentence-transformers`?

Personally, I prefer `conda` to `pip` whenever possible. Could you perhaps update the `conda` index or associated dependency tree for `sentence-transformers`. I don't actually really get how all that works.

Regardless, I vote to close this issue if you feel it's been resolved.

**Comment by paulpacaud** [comment]
Same issue here:
ImportError: cannot import name 'is_torch_npu_available' from 'transformers' (unknown location)

The problem does not seem to be solved

**Comment by tomaarsen** [comment]
Can you please add
```python
from transformers import __version__ as transformers_version
print(transformers_version)
```
at the start of your script? I suspect that you might be installing into a different virtual environment than the one that you're running from.

- Tom Aarsen
